generate inplace/out kernels for xla

bdhirsh opened the pull request:
This is a re-write of https://github.com/pytorch/pytorch/pull/56835, which is significantly shorter thanks to the data model change in the PR below this one in the stack. See the original description in the linked PR for details.

The functional changes in this PR are the same as in the above linked one, so the description is the same with a few small changes:
- I don't bother generating `at::xla::{op}` entries for CPU fallbacks. After looking around, I see precedent for that. For example, we don't have `at::cpu::{op}` entries for composite ops- if you really want to bypass the dispatcher you need to call `at::compositeimplicitautograd::{op}`. Maybe we should revisit that later if we find an important use case for having full namespace coverage, but that doesn't seem worth half-fixing for external backends in this PR.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58065 add a boxed CPU fallback kernel
* #58064 remove xla-specific stuff from codegen (minus CPU fallback)
* **#57510 generate inplace/out kernels for xla**
* #57361 [codegen] split out backend-specific information from NativeFunction in the model
* #55796 remove bridge API from codegen
* #56597 [external codegen] better yaml error messaging, added explicit error message tests
* #55795 add _to_cpu() operator

Differential Revision: [D28474364](https://our.internmc.facebook.com/intern/diff/D28474364)

bdhirsh committed:
generate inplace/out kernels for xla

[ghstack-poisoned]

bdhirsh committed:
Update on "generate inplace/out kernels for xla"




[ghstack-poisoned]

bdhirsh committed:
Update on "generate inplace/out kernels for xla"




[ghstack-poisoned]

bdhirsh commented on tools/codegen/dest/register_dispatch_key.py:
```diff
@@ -103,109 +175,115 @@ def gen_structured(self, g: NativeFunctionsGroup) -> List[str]:
-                inplace_meta = True
-            else:
+    def gen_unstructured(self, f: NativeFunction, *, is_generated_wrapper: bool = False) -> Optional[str]:
+        with native_function_manager(f):
```
The `with native_function_manager(f)` makes the diff painful, but there aren't too many changes in this function. As per the comment [in the last PR](https://github.com/pytorch/pytorch/pull/56835#discussion_r620822511), I can't easily use a decorator since I have a defaultable argument

ezyang replied:
It's fine, reviewers can ignore whitespace in GH UI

bdhirsh commented:
Hooking this up to an XLA-side PR to sanity check.

bdhirsh committed:
Update on "generate inplace/out kernels for xla"

This is a re-write of https://github.com/pytorch/pytorch/pull/56835, which is significantly shorter thanks to the data model change in the PR below this one in the stack. See the original description in the linked PR for details.

The functional changes in this PR are the same as in the above linked one, so the description is the same with a few small changes:
- I don't bother generating `at::xla::{op}` entries for CPU fallbacks. After looking around, I see precedent for that. For example, we don't have `at::cpu::{op}` entries for composite ops- if you really want to bypass the dispatcher you need to call `at::compositeimplicitautograd::{op}`. Maybe we should revisit that later if we find an important use case for having full namespace coverage, but that doesn't seem worth half-fixing for external backends in this PR.




[ghstack-poisoned]

ezyang commented on tools/codegen/model.py:
```diff
@@ -493,9 +493,14 @@ def __post_init__(self) -> None:
-    def functions(self) -> Iterator[NativeFunction]:
-        yield self.out
-        yield self.functional
+    def functions(self, *, functional_first: bool = False) -> Iterator[NativeFunction]:
```
Kind of goofy

bdhirsh replied:
yeah you're right, I think we should be able to unconditionally use the functional kernel first.

ezyang commented on tools/codegen/dest/register_dispatch_key.py:
```diff
@@ -312,14 +390,14 @@ def gen_class_set_output_body(self, k: SchemaKind) -> str:
 TORCH_CHECK(options.device() == out.device(),
     "Expected out tensor to have device ", options.device(), ", but got ", out.device(), " instead");
-bool resized = at::native::resize_output(outputs_[output_idx], sizes);
+bool resized = {self.cpp_namespace}::resize_output(outputs_[output_idx], sizes);
```
This seems a little questionable; are you actually changing what namespace we reference `resize_output` here?

bdhirsh replied:
Yep that's a bug, thanks! Since `at::native::resize_output` is the thing we want to call to resize the output tensor, regardless of which backend we're generating. We wouldn't have hit it until we eventually tried to enable structured external kernels.

ezyang replied:
Also for the empty calls above, I think

ezyang commented on tools/codegen/dest/gen_external_aten_fallbacks.py:
```diff
@@ -194,24 +156,18 @@ def get_device_param(args: List[Argument]) -> str:
+                # We also skip registrations if there is a functional backend kernel,
+                # because we generate out/inplace wrappers in that case (handled in register_dispatch_key.py).
+                if self.backend_index.get(f) is not None or \
+                        (isinstance(g, NativeFunctionsGroup) and self.backend_index.get(g.functional) is not None):
```
The downside to having this split in a separate dest here is that you have to keep the logic here in sync with register_dispatch_key.py for when to generate out/inplace wrappers. This is probably... OK, because there are fairly large benefits to having this be a separate file, but it's probably good to explicitly cross ref the corresponding condition and/or have a single function where you do the test that commons up the logic (even though it is short to type).

ezyang commented on tools/codegen/dest/register_dispatch_key.py:
```diff
@@ -85,6 +98,65 @@ def __call__(self, f: Union[NativeFunctionsGroup, NativeFunction]) -> List[str]:
         else:
             assert_never(f)
 
+    def gets_generated_out_inplace_wrapper(self, f: NativeFunction, g: NativeFunctionsGroup) -> bool:
```
This is the code to xref, I think

bdhirsh replied:
yep thanks, this should be shared in `gen_external_aten_fallbacks.py`. Another thing to not worry about later when/if the boxed CPU fallback works out. (That's also why I kind of wanted to keep all the fallback stuff in a separate file, so it'll be easier to disentangle if we want to)

ezyang commented on tools/codegen/dest/register_dispatch_key.py:
```diff
@@ -85,6 +98,65 @@ def __call__(self, f: Union[NativeFunctionsGroup, NativeFunction]) -> List[str]:
+            updates = f'at::_copy_from_and_resize({functional_result_name}, {ret_name});'
+            returns = ret_name
+
+        functional_sig = self.wrapper_kernel_sig(g.functional)
```
Although you can skip the translate call here because everything is the same, I wonder if it would be clearer for readers to translate anyway... (maybe there's some stuff translate chokes on?)

ezyang reviewed (approved):
please check reviews

ezyang commented on tools/codegen/dest/register_dispatch_key.py:
```diff
@@ -85,6 +98,65 @@ def __call__(self, f: Union[NativeFunctionsGroup, NativeFunction]) -> List[str]:
+        # All other functionality can be directly pulled from gen_unstructured.
+        is_generated_wrapper = self.gets_generated_out_inplace_wrapper(f, g)
+        if self.target is not Target.ANONYMOUS_DEFINITION:
+            return self.gen_unstructured(f, is_generated_wrapper=is_generated_wrapper)
```
Might be a good idea to just pass it unconditionally, tbh, in case more codegen in unstructured kernels start making use of the group information.

bdhirsh commented on tools/codegen/dest/register_dispatch_key.py:
```diff
@@ -85,6 +98,65 @@ def __call__(self, f: Union[NativeFunctionsGroup, NativeFunction]) -> List[str]:
+        return_names = cpp.return_names(f)
+        if len(return_names) > 1:
+            updates = '\n  '.join(
+                f'at::_copy_from_and_resize(std::get<{i}>({functional_result_name}), {ret_name});'
```
Oh, good catch. I think that any size mismatch errors would be caught in the meta function, so we should never end up in that situation. But getting the codegen to explicitly call a non-resizing function for inplace wrappers definitely expresses intent more clearly.

bdhirsh replied:
Actually yeah that was a straight up bug- since a lot of XLA kernels don't do shape checks (LazyTensor `meta()` support can't come soon enough). So before when I was using `_copy_from_and_resize()` for inplace ops, we would just re-size `self` on XLA, while CPU would correctly error in the meta function.

using `_copy_from` gives you a pretty crappy error message, but there's probably no point trying to make that better until we start using meta functions.

bdhirsh committed:
Update on "generate inplace/out kernels for xla"

This is a re-write of https://github.com/pytorch/pytorch/pull/56835, which is significantly shorter thanks to the data model change in the PR below this one in the stack. See the original description in the linked PR for details.

The functional changes in this PR are the same as in the above linked one, so the description is the same with a few small changes:
- I don't bother generating `at::xla::{op}` entries for CPU fallbacks. After looking around, I see precedent for that. For example, we don't have `at::cpu::{op}` entries for composite ops- if you really want to bypass the dispatcher you need to call `at::compositeimplicitautograd::{op}`. Maybe we should revisit that later if we find an important use case for having full namespace coverage, but that doesn't seem worth half-fixing for external backends in this PR.




[ghstack-poisoned]

bhosmer commented on tools/codegen/api/types.py:
```diff
@@ -407,11 +407,13 @@ class DispatcherSignature:
     # The schema this signature is derived from
     func: FunctionSchema
 
+    prefix: str = ""
```
worth a comment? it opens up a pretty big axis of variation for what the class represents...

bdhirsh replied:
Definitely fair, I'll add it in

bhosmer commented on tools/codegen/api/types.py:
```diff
@@ -441,8 +443,8 @@ def type(self) -> str:
     @staticmethod
-    def from_schema(func: FunctionSchema) -> 'DispatcherSignature':
-        return DispatcherSignature(func)
+    def from_schema(func: FunctionSchema, *, prefix: str = '') -> 'DispatcherSignature':
```
ditto re comment

bhosmer commented on tools/codegen/api/types.py:
```diff
@@ -483,14 +485,15 @@ def dispatcher_exprs(self) -> List[Expr]:
 # Helper functions
 
-def kernel_signature(f: NativeFunction, backend_index: BackendIndex) -> Union['NativeSignature', 'DispatcherSignature']:
+def kernel_signature(
```
ditto re comment for `prefix` - also I'm not far enough into the PR yet to understand why it's a freestanding param and not part of the data model.

bhosmer commented on tools/codegen/dest/register_dispatch_key.py:
```diff
@@ -85,6 +92,51 @@ def __call__(self, f: Union[NativeFunctionsGroup, NativeFunction]) -> List[str]:
 
+    def wrapper_kernel_sig(self, f: NativeFunction) -> Union[NativeSignature, DispatcherSignature]:
+        # The prefix is just to ensure uniqueness. The Dispatcher API doesn't guarantee unique kernel names.
+        return kernel_signature(f, self.backend_index, prefix=f'wrapper_{f.func.name.overload_name}_')
```
Just a high level observation and not to overengineer, but here `prefix` is derived from `f` in a pretty straightforward way. If that's true everywhere, then it might be cleaner to host the prefix-making logic in `NativeFunction`.

bdhirsh replied:
So, let me know if we disagree, but the value of `prefix` here seems more to me like an implementation detail of how we code-generate the stuff that we register to the dispatcher than a property of a NativeFunction. Rather than directly registering native kernels to the dispatcher, the codegen creates anonymous wrappers that call into to the native kernels, and registers those wrappers instead. We just append `wrapper_` to each wrapper function's name for clarity, and to avoid name collisions with the namespaced definitions that we generate in the same file.

So I kind of see `NativeSignature`/`DispatcherSignature` as a lens into how to display a given NativeFunction's signature, like what schema and naming transformations to apply, and the wrapper prefix is just another knob along those same lines.

bhosmer replied:
Ah sorry! Not `NativeFunction` but rather `NativeSignature` and `DispatchSignature`. You're right that prefixes shouldn't be part of `NativeFunction`'s world. But the only "free" value here is the 'wrapper_{}_' - the function name in the middle comes from `f`, which is 1-1 with whichever Signature `wrapper_kernel_sig()` constructs - so making the caller supply the correspondence here would only be useful if we ever wanted to break that relationship, which I don't think we would, right? It seems more like an opportunity for the caller to mess things up by passing a constant string or whatever.

Anyway to cut to the chase, the whole correct by construction thing would say, only pass the constant part of the prefix here and onward into the Signatures, and make them (the Signatures) responsible for folding `f.func.name.overload_name` into nonempty prefixes. 

(BTW if you don't actually use multiple prefixes, you could just have the Signatures carry a bit saying whether to mangle in the overload name. (And if `kernel_signature()` is always called with a nonempty prefix, you could just have it unilaterally construct the Signature instances with prefixes turned on.)

bdhirsh committed:
Update on "generate inplace/out kernels for xla"

This is a re-write of https://github.com/pytorch/pytorch/pull/56835, which is significantly shorter thanks to the data model change in the PR below this one in the stack. See the original description in the linked PR for details.

The functional changes in this PR are the same as in the above linked one, so the description is the same with a few small changes:
- I don't bother generating `at::xla::{op}` entries for CPU fallbacks. After looking around, I see precedent for that. For example, we don't have `at::cpu::{op}` entries for composite ops- if you really want to bypass the dispatcher you need to call `at::compositeimplicitautograd::{op}`. Maybe we should revisit that later if we find an important use case for having full namespace coverage, but that doesn't seem worth half-fixing for external backends in this PR.




[ghstack-poisoned]

bdhirsh committed:
Update on "generate inplace/out kernels for xla"

This is a re-write of https://github.com/pytorch/pytorch/pull/56835, which is significantly shorter thanks to the data model change in the PR below this one in the stack. See the original description in the linked PR for details.

The functional changes in this PR are the same as in the above linked one, so the description is the same with a few small changes:
- I don't bother generating `at::xla::{op}` entries for CPU fallbacks. After looking around, I see precedent for that. For example, we don't have `at::cpu::{op}` entries for composite ops- if you really want to bypass the dispatcher you need to call `at::compositeimplicitautograd::{op}`. Maybe we should revisit that later if we find an important use case for having full namespace coverage, but that doesn't seem worth half-fixing for external backends in this PR.




[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28474364).
